test/performance/odp_bench_event_vector.c

```diff
@@ -0,0 +1,760 @@
+#define TEST_ROUNDS 100u
+
+/** Maximum number of results to be held */
+#define TEST_MAX_BENCH 20
+
+#define BENCH_INFO(run_fn, init_fn, term_fn, alt_name) \
+	{.name = #run_fn, .run = run_fn, .init = init_fn, .term = term_fn, .desc = alt_name}
+
+/**
+ * Parsed command line arguments
+ */
+typedef struct {
+	int bench_idx;   /** Benchmark index to run indefinitely */
+	int cache_size;  /** Pool cache size */
+	int time;        /** Measure time vs. CPU cycles */
+	uint32_t rounds; /** Rounds per test case */
+	uint32_t size;   /** Number of event in each event vector */
+} appl_args_t;
+
+/**
+ * Grouping of all global data
+ */
+typedef struct {
+	/** Application (parsed) arguments */
+	appl_args_t appl;
+	/** Common benchmark suite data */
+	bench_suite_t suite;
+	/** Event vector pool */
+	odp_pool_t evv_pool;
+	/** Buffer pool */
+	odp_pool_t buf_pool;
+	/** Number of events in event vector */
+	uint32_t evv_size;
+	/** Size of buffer */
+	uint32_t buf_size;
+	/** Event vector user area size */
+	uint32_t evv_uarea_size;
+	/** Buffer user area size */
+	uint32_t buf_uarea_size;
+	/** Array for storing test event vectors */
+	odp_event_vector_t evv_arr[TEST_REPEAT_COUNT];
+	/** Array for storing test event */
+	odp_event_t event_tbl[TEST_REPEAT_COUNT];
+	/** Array for storing event vector table pointers */
+	odp_event_t *evv_tbl[TEST_REPEAT_COUNT][TEST_MAX_EVV_SIZE];
+	/** Array for storing test pointers */
+	void *ptr_tbl[TEST_REPEAT_COUNT];
+	/** Array for storing test pool handles */
+	odp_pool_t pool_tbl[TEST_REPEAT_COUNT];
+	/** Array for storing test event types */
+	odp_event_type_t event_type_tbl[TEST_REPEAT_COUNT];
+	/** CPU mask as string */
+	char cpumask_str[ODP_CPUMASK_STR_SIZE];
+	/** Array for storing results */
+	double result[TEST_MAX_BENCH];
+} args_t;
+
+/* Global pointer to args */
+static args_t *gbl_args;
+
+static void sig_handler(int signo ODP_UNUSED)
+{
+	if (gbl_args == NULL)
+		return;
+	odp_atomic_store_u32(&gbl_args->suite.exit_worker, 1);
+}
+
+static int setup_sig_handler(void)
+{
+	struct sigaction action = { .sa_handler = sig_handler };
+
+	if (sigemptyset(&action.sa_mask) || sigaction(SIGINT, &action, NULL))
+		return -1;
+
+	return 0;
+}
+
+static void allocate_test_evv_and_contents(odp_event_vector_t evv[], int num)
+{
+	odp_pool_t evv_pool = gbl_args->evv_pool;
+	odp_pool_t buf_pool = gbl_args->buf_pool;
+	uint32_t evv_size = gbl_args->evv_size;
+	odp_buffer_t buf[evv_size];
+	odp_event_t *tbl;
+
+	for (int i = 0; i < num; i++) {
+		evv[i] = odp_event_vector_alloc(evv_pool);
+		if (evv[i] == ODP_EVENT_VECTOR_INVALID)
+			ODPH_ABORT("Allocating test event vector failed\n");
+
+		odp_event_vector_tbl(evv[i], &tbl);
+		if (odp_buffer_alloc_multi(buf_pool, buf, evv_size) != (int)evv_size) {
+			for (uint32_t j = 0; j < evv_size; j++) {
+				if (buf[j] == ODP_BUFFER_INVALID)
+					ODPH_ABORT("Alloc failed %u/%u\n", i, j);
+			}
+			ODPH_ABORT("Not enough buffers allocated!\n");
+		}
```

Comment:
Depending on the implementation, `odp_buffer_alloc_multi()` may return less than the requested number of buffers per call and only return value <0 should be considered a failure. So, a retry loop is required here. Also, no need to check the individual buffers after alloc.